platform/rct/hal/usart_stm32/usart_device_stm32.c

```diff
@@ -454,3 +481,249 @@ void USART3_IRQHandler(void)
+        }
+
+        ui->dma_rx.ptr = head;
+        ui->rx_overflow |= overflow;
+
+        return task_awoke;
+}
+
+void DMA1_Channel5_IRQHandler(void)
+{
+        volatile struct usart_info *ui = usart_data + UART_WIRELESS;
+        const bool task_awoken = dma_rx_to_queue_isr(ui);
+        DMA_ClearITPendingBit(DMA1_IT_TC5);
+        DMA_ClearITPendingBit(DMA1_IT_HT5);
+        portEND_SWITCHING_ISR(task_awoken);
+}
+
+void DMA1_Channel6_IRQHandler(void)
+{
+        volatile struct usart_info *ui = usart_data + UART_GPS;
+        const bool task_awoken = dma_rx_to_queue_isr(ui);
+        DMA_ClearITPendingBit(DMA1_IT_TC6);
+        DMA_ClearITPendingBit(DMA1_IT_HT6);
+        portEND_SWITCHING_ISR(task_awoken);
+}
+
+static bool complete_dma_tx_isr(volatile struct usart_info* ui)
+{
+        volatile uint8_t* buff = ui->dma_tx.buff;
+        volatile uint8_t* const edge = buff + ui->dma_tx.buff_size;
+        volatile uint8_t* head = buff;
+        xQueueHandle queue = serial_get_tx_queue(ui->serial);
+        portBASE_TYPE task_awoke = pdFALSE;
+
+        DMA_Cmd(ui->dma_tx.chan, DISABLE);
+
+        /*
+         * Now that we are done transferring, see if there is anything
+         * else in the Tx queue that needs to be sent. If so, copy it
+         * into the buffer.  If we fill up more than 1/2 of the buffer,
+         * then lets start another transfer. Else defer until the next
+         * timer interrupt. This helps ensure that we limit the # of
+         * interrupts we need to service at the expense of a tiny bit
+         * of latency.
+         */
+        for (uint8_t var; head < edge; ++head) {
+                if (!xQueueReceiveFromISR(queue, &var, &task_awoke))
+                        break;
+
+                *head = var;
+        }
+
+        const size_t min_tx_size = ui->dma_tx.buff_size / 2;
+        const size_t curr_tx_size = (size_t) (head - buff);
+        if (min_tx_size < curr_tx_size) {
+                /* Then we have enough to transfer */
+                ui->dma_tx.chan->CNDTR = (uint32_t) curr_tx_size;
+                DMA_Cmd(ui->dma_tx.chan, ENABLE);
+        } else {
+                /* Not enough. Set ptr to head to indicate DMA done. */
+                ui->dma_tx.ptr = head;
+        }
+
+        return task_awoke;
+}
+
+void DMA1_Channel4_IRQHandler(void)
+{
+        volatile struct usart_info *ui = usart_data + UART_WIRELESS;
+        const bool task_awoken = complete_dma_tx_isr(ui);
+        DMA_ClearITPendingBit(DMA1_IT_TC4);
+        portEND_SWITCHING_ISR(task_awoken);
+}
+
+void DMA1_Channel7_IRQHandler(void)
+{
+        volatile struct usart_info *ui = usart_data + UART_GPS;
+        const bool task_awoken = complete_dma_tx_isr(ui);
+        DMA_ClearITPendingBit(DMA1_IT_TC7);
+        portEND_SWITCHING_ISR(task_awoken);
+}
+
+
+/* *** Timer Methods *** */
+static void dma_rx_to_queue(volatile struct usart_info *ui)
+{
+        taskENTER_CRITICAL();
```

Comment:
Yes.  Can't have the task be interrupted by a different interrupt (like the DMA bit itself) or the system itself since doing so would invalidate the pointer calculations I make at the beginning.